File: apps/api/app/routers/resources.py
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, Query, Response
from sqlalchemy import delete, select
from sqlalchemy.orm import Session

from app.db import get_db
from app.models import Language, Resource, Topic
from app.schemas import ResourceIn, ResourceOut, iso

router = APIRouter(tags=["resources"])


def _resource_out(row: dict) -> ResourceOut:
    return ResourceOut(
        id=row["id"],
        title=row["title"],
        url=row["url"],
        type=row["type"],
        description=row.get("description"),
        language_id=row.get("language_id"),
        language_name=row.get("language_name"),
        topic_id=row.get("topic_id"),
        topic_title=row.get("topic_title"),
        tags=row.get("tags"),
        created_at=iso(row.get("created_at")),
    )


@router.get("/resources", response_model=list[ResourceOut])
def list_resources(
    language_id: int | None = Query(None, alias="languageId"),
    topic_id: int | None = Query(None, alias="topicId"),
    type: str | None = None,
    db: Session = Depends(get_db),
) -> list[ResourceOut]:
    stmt = (
        select(
            Resource.id,
            Resource.title,
            Resource.url,
            Resource.type,
            Resource.description,
            Resource.language_id,
            Resource.topic_id,
            Resource.tags,
            Resource.created_at,
            Language.name.label("language_name"),
            Topic.title.label("topic_title"),
        )
        .outerjoin(Language, Language.id == Resource.language_id)
        .outerjoin(Topic, Topic.id == Resource.topic_id)
        .order_by(Resource.id)
    )
    if language_id is not None:
        stmt = stmt.where(Resource.language_id == language_id)
    if topic_id is not None:
        stmt = stmt.where(Resource.topic_id == topic_id)
    if type:
        stmt = stmt.where(Resource.type == type)
    rows = db.execute(stmt).mappings().all()
    return [_resource_out(dict(r)) for r in rows]


@router.post("/resources", response_model=ResourceOut, status_code=201)
def create_resource(body: ResourceIn, db: Session = Depends(get_db)) -> ResourceOut:
    resource = Resource(
        title=body.title,
        url=body.url,
        type=body.type,
        description=body.description,
        language_id=body.language_id,
        topic_id=body.topic_id,
        tags=body.tags,
    )
    db.add(resource)
    db.commit()
    db.refresh(resource)
    return ResourceOut(
        id=resource.id,
        title=resource.title,
        url=resource.url,
        type=resource.type,
        description=resource.description,
        language_id=resource.language_id,
        topic_id=resource.topic_id,
        tags=resource.tags,
        created_at=iso(resource.created_at),
    )


@router.get("/resources/{resource_id}", response_model=ResourceOut)
def get_resource(resource_id: int, db: Session = Depends(get_db)) -> ResourceOut:
    stmt = (
        select(
            Resource.id,
            Resource.title,
            Resource.url,
            Resource.type,
            Resource.description,
            Resource.language_id,
            Resource.topic_id,
            Resource.tags,
            Resource.created_at,
            Language.name.label("language_name"),
            Topic.title.label("topic_title"),
        )
        .outerjoin(Language, Language.id == Resource.language_id)
        .outerjoin(Topic, Topic.id == Resource.topic_id)
        .where(Resource.id == resource_id)
    )
    row = db.execute(stmt).mappings().first()
    if not row:
        raise HTTPException(404, "Resource not found")
    return _resource_out(dict(row))


@router.delete("/resources/{resource_id}", status_code=204)
def delete_resource(resource_id: int, db: Session = Depends(get_db)) -> Response:
    result = db.execute(delete(Resource).where(Resource.id == resource_id))
    if result.rowcount == 0:
        raise HTTPException(404, "Resource not found")
    db.commit()
    return Response(status_code=204)

File: apps/api/src/routes/resources.ts
import { Router } from "express";
import { and, eq } from "drizzle-orm";
import {
  db,
  resourcesTable,
  languagesTable,
  topicsTable,
  insertResourceSchema,
} from "@devlearn/database";
import { CreateResourceBody } from "@devlearn/api-zod";
import { asyncHandler } from "../lib/async.ts";
import { HttpError } from "../lib/errors.ts";
import { parseIntParam, parseOptionalInt, parseOptionalString } from "../lib/parse.ts";

export const resourcesRouter = Router();

const resourceSelect = {
  id: resourcesTable.id,
  title: resourcesTable.title,
  url: resourcesTable.url,
  type: resourcesTable.type,
  description: resourcesTable.description,
  languageId: resourcesTable.languageId,
  topicId: resourcesTable.topicId,
  tags: resourcesTable.tags,
  createdAt: resourcesTable.createdAt,
  languageName: languagesTable.name,
  topicTitle: topicsTable.title,
};

function serializeResource(r: {
  createdAt: Date | null;
  [k: string]: unknown;
}) {
  return { ...r, createdAt: r.createdAt ? r.createdAt.toISOString() : null };
}

resourcesRouter.get(
  "/resources",
  asyncHandler(async (req, res) => {
    const languageId = parseOptionalInt(req.query.languageId);
    const topicId = parseOptionalInt(req.query.topicId);
    const type = parseOptionalString(req.query.type);

    const where = [];
    if (languageId !== undefined) where.push(eq(resourcesTable.languageId, languageId));
    if (topicId !== undefined) where.push(eq(resourcesTable.topicId, topicId));
    if (type) where.push(eq(resourcesTable.type, type));

    const rows = await db
      .select(resourceSelect)
      .from(resourcesTable)
      .leftJoin(languagesTable, eq(languagesTable.id, resourcesTable.languageId))
      .leftJoin(topicsTable, eq(topicsTable.id, resourcesTable.topicId))
      .where(where.length ? and(...where) : undefined)
      .orderBy(resourcesTable.id);

    res.json(rows.map(serializeResource));
  }),
);

resourcesRouter.post(
  "/resources",
  asyncHandler(async (req, res) => {
    const body = CreateResourceBody.parse(req.body);
    const input = insertResourceSchema.parse(body);
    const [created] = await db.insert(resourcesTable).values(input).returning();
    if (!created) throw new HttpError(500, "Failed to create resource");
    res.status(201).json(serializeResource(created));
  }),
);

resourcesRouter.get(
  "/resources/:id",
  asyncHandler(async (req, res) => {
    const id = parseIntParam(req.params.id, "id");
    const [row] = await db
      .select(resourceSelect)
      .from(resourcesTable)
      .leftJoin(languagesTable, eq(languagesTable.id, resourcesTable.languageId))
      .leftJoin(topicsTable, eq(topicsTable.id, resourcesTable.topicId))
      .where(eq(resourcesTable.id, id));
    if (!row) throw new HttpError(404, "Resource not found");
    res.json(serializeResource(row));
  }),
);

resourcesRouter.delete(
  "/resources/:id",
  asyncHandler(async (req, res) => {
    const id = parseIntParam(req.params.id, "id");
    const result = await db
      .delete(resourcesTable)
      .where(eq(resourcesTable.id, id))
      .returning({ id: resourcesTable.id });
    if (result.length === 0) throw new HttpError(404, "Resource not found");
    res.status(204).end();
  }),
);
